src/main/java/spoon/reflect/visitor/chain/CtQuery.java

```diff
@@ -23,59 +23,91 @@
- * <li> {@link #filterChildren(Filter))} - uses {@link Filter} instances to filter children of an element
- * <li> {@link #list()} - to evaluate the query and return a list of elements produced by this query.
- * </ul>
- * It makes sense to evaluate this query only once, because the input element is constant.<br>
- * A CtQuery is lazily evaluated once {{@link #list()}} or {@link #forEach(CtConsumer)} are called.
- * Usually a new query is created each time when one needs to query something.
- * If you need to reuse a query instance several times, for example in a loop, then use {@link CtQuery#setInput(Object...)}
- * to bound this query with different input.
+ * Chaining: In a query several steps can be chained, by chaining calls to map functions. The non-null outputs of one step
+ * are given as input to the next step. An iterable or array output is considered as a set of different inputs for the next step.
+ *
+ * Evaluation: A CtQuery is lazily evaluated once {{@link #list()}} or {@link #forEach(CtConsumer)} are called.<br/>
  *
  */
 public interface CtQuery extends CtQueryable {
 
 	/**
-	 * sets (binds) the input of the query. If the query is created by {@link CtElement#map} or {@link CtElement#filterChildren(Filter)},
+	 * Recursively scans all children elements of an input element.
+	 * The matched child element for which (filter.matches(element)==true) are sent to the next query step.
+	 * Essentially the same as {@link CtElement#getElements(Filter)} but more powerful, because it
+	 * can be chained with other subsequent queries.
+	 *
+	 * Note: the input element (the root of the query, `this` if you're in {@link CtElement}) is also checked and may thus be also sent to the next step.
+	 * The elements which throw {@link ClassCastException} during {@link Filter#matches(CtElement)}
+	 * are considered as **not matching**, ie. are excluded.
+	 *
+	 * @param filter used to filter scanned children elements of the AST tree
+	 * @return a new Query
+	 */
+	@Override
+	<R extends CtElement> CtQuery filterChildren(Filter<R> filter);
+
+
+	/**
+	 * Query elements based on a function, the behavior depends on the return type of the function.
+	 * <table>
+	 * <tr><td><b>Return type of `function`</b><td><b>Behavior</b>
+	 * <tr><td>{@link Boolean}<td>Select elements if thereturned value of `function` is true (as for {@link Filter}).
+	 * <tr><td>? extends {@link Object}<td>Send the returned value of `function` to the next step
+	 * <tr><td>{@link Iterable}<td>Send each item of the collection to the next step
+	 * <tr><td>{@link Object[]}<td>Send each item of the array to the next step
+	 * </table><br>
+	 *
+	 * @param function a Function with one parameter of type I returning a value of type R
+	 * @return a new query object
+	 */
+	@Override
+	<I, R> CtQuery map(CtFunction<I, R> function);
+
+	/**
+	 * Sets (binds) the input of the query, the search will be done on children of that element.

```

Comment:
the search will be done **on children** of that element - it is true only for filterChildren... but how to write it better? :-)